mgr/dashboard: RBD management (backend)

rjfd opened the pull request:
This PR adds the RBD management functionality in the dashboard backend.

Signed-off-by: Ricardo Dias <rdias@suse.com>

LenzGr commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -1,136 +1,421 @@
+    rbd.RBD_FEATURE_OPERATIONS: "operations",
+}
+
+def _format_bitmask(features):
```
The linter rightfully complained here: `./controllers/rbd.py:57:1: E302 expected 2 blank lines, found 1`

rjfd commented:
QA tests run successfuly: http://pulpito.ceph.com/rdias-2018-04-12_12:58:27-rados:mgr-wip-rdias-testing-distro-basic-smithi/

dillaman commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -112,6 +136,25 @@ def _rbd_image(self, ioctx, pool_name, image_name):
             stat['snapshots'].append(snap)
 
+        # disk usage
+        if 'fast-diff' in stat['features_name']:
```
Nit: note that this isn't a "free" operation even w/ fast-diff enabled. For each snapshot, it requires loading the object map for said snapshot from the OSDs to compute the usage. If this image metadata is cached / loading asynchronously in the background, it probably isn't too big of a deal. But if this will stall the UI while it's opening lots of images and computing usage, that might be sub-optimal.

rjfd replied:
Thanks for the explanation. Yes, this code runs asynchronously and its results are cached.

rjfd replied:
ah, you're right. I was trying to calculate something that only a filesystem stored inside image actually knows. Forget about that.

dillaman commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -249,6 +257,72 @@ def delete(self, pool_name, image_name):
+        return {'success': True}
+
+    @RESTController.args_from_json
+    def set(self, pool_name, image_name, name=None, size=None, features=None):
```
Perhaps support flatten?

ricardoasmarques commented:
FYI, the frontend part of this PR is available on https://github.com/ceph/ceph/pull/21385

dillaman commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -1,136 +1,422 @@
+            stat['snapshots'].append(snap)
+
+        # disk usage
+        if 'fast-diff' in stat['features_name']:
```
... should also verify that the fast-diff isn't invalid via ```flags()``` 

[1] https://github.com/ceph/ceph/blob/master/src/pybind/rbd/rbd.pyx#L78

dillaman commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -1,136 +1,422 @@
+        return _rbd_image_call(pool_name, image_name, _src_copy)
+
+
+@ApiController('rbd/:pool_name/:image_name/snap')
```
Nit: why not have all block-related things under a ```block``` sub-directory? iSCSI is ```block/iscsi```, this could be ```block/images/...```, and mirroring could be ```block/mirroring```.

rjfd commented:
@dillaman one of the things I missed from python librbd API is the async versions of some operations, like "remove", or "rollback". Is there any technical challenge for providing the async versions of those methods, or it is just a matter of no one worked on them?

I will probably try to add these async versions in the future, so that we can show the progress of such operations in the dashboard UI.

sebastian-philipp commented:
You could try to improve the JSON validation by something like
```diff
git diff qa/tasks/mgr/dashboard/test_rbd.py | cat
diff --git a/qa/tasks/mgr/dashboard/test_rbd.py b/qa/tasks/mgr/dashboard/test_rbd.py
index ecd8d40c10..b119f4f5ba 100644
--- a/qa/tasks/mgr/dashboard/test_rbd.py
+++ b/qa/tasks/mgr/dashboard/test_rbd.py
@@ -118,24 +118,28 @@ class RbdTest(DashboardTestCase):
                               "object-map"]
         }
         """
-        self.assertIn('size', img)
-        self.assertIn('obj_size', img)
-        self.assertIn('num_objs', img)
-        self.assertIn('order', img)
-        self.assertIn('block_name_prefix', img)
-        self.assertIn('name', img)
-        self.assertIn('id', img)
-        self.assertIn('pool_name', img)
-        self.assertIn('features', img)
-        self.assertIn('features_name', img)
-        self.assertIn('stripe_count', img)
-        self.assertIn('stripe_unit', img)
-        self.assertIn('parent', img)
-        self.assertIn('data_pool', img)
-        self.assertIn('snapshots', img)
-        self.assertIn('timestamp', img)
-        self.assertIn('disk_usage', img)
-        self.assertIn('total_disk_usage', img)
+        from .helper import JObj, JLeaf, JList, JAny
+        schema = JObj(sub_elems={
+            'size': JLeaf(int),
+            'obj_size': JLeaf(int),
+            'num_objs': JLeaf(int),
+            'order': JLeaf(int),
+            'block_name_prefix': JLeaf(str),
+            'name': JLeaf(str),
+            'id': JLeaf(str),
+            'pool_name': JLeaf(str),
+            'features': JLeaf(int),
+            'features_name': JList(JLeaf(str)),
+            'stripe_count': JLeaf(int, none=True),
+            'stripe_unit': JLeaf(int, none=True),
+            'parent': JLeaf(str, none=True),
+            'data_pool': JLeaf(str, none=True),
+            'snapshots': JAny(),
+            'timestamp': JLeaf(str, none=True),
+            'disk_usage': JLeaf(str, none=True),
+            'total_disk_usage': JLeaf(str, none=True),
+        })
+        self.assertSchema(img, schema)
 
         for k, v in kwargs.items():
             if isinstance(v, list):
```

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -294,6 +363,7 @@ def _parse_function_args(cls, func):
                      (v.kind == inspect.Parameter.POSITIONAL_ONLY or
                       v.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD)]
         else:
+            func = getattr(func, '__wrapped__', func)
```
AFAIK, this only works in Python 3. I'm working on a solution that also works for python2

rjfd replied:
this also works in python 2 because I'm explicitly adding that attribute in: https://github.com/ceph/ceph/pull/21360/files#diff-3095fd00dfad68aeeccd8ad3fe6d3356R327

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -1,136 +1,431 @@
+    return _rbd_call(pool_name, _ioctx_func, image_name, func, *args, **kwargs)
+
+
+RBD_FEATURES_NAME_MAPPING = {
```
How can we make sure, this is not getting out-of-date?

rjfd replied:
don't know, we should figure out a better way in the future.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/rbd.py:
```diff
@@ -1,136 +1,431 @@
-            features = i.features()
+    # set of image features that can be disabled on existing images
+    ALLOW_DISABLE_FEATURES = set(["exclusive-lock", "object-map", "fast-diff",
+                                  "deep-flatten", "journaling"])
```
How can we make sure, this is not getting out-of-date?

rjfd replied:
Update the code when someone changes the Ceph RBD component. :-)

dillaman replied:
We can expose the RBD_FEATURES_MUTABLE bitmasks via the RBD python module

s0nea commented:
I'm receiving the following failure when running './run-backend-api-tests.sh tasks.mgr.dashboard.test_rbd.RbdTest' on my local docker environment:
```
2018-04-13 10:50:16,370.370 INFO:__main__:======================================================================
2018-04-13 10:50:16,371.371 INFO:__main__:FAIL: test_clone (tasks.mgr.dashboard.test_rbd.RbdTest)
2018-04-13 10:50:16,371.371 INFO:__main__:----------------------------------------------------------------------
2018-04-13 10:50:16,371.371 INFO:__main__:Traceback (most recent call last):
2018-04-13 10:50:16,371.371 INFO:__main__:  File "/ceph/qa/tasks/mgr/dashboard/test_rbd.py", line 463, in test_clone
2018-04-13 10:50:16,371.371 INFO:__main__:    'snap_name': 'snap1'})
2018-04-13 10:50:16,371.371 INFO:__main__:  File "/ceph/qa/tasks/mgr/dashboard/test_rbd.py", line 137, in _validate_image
2018-04-13 10:50:16,371.371 INFO:__main__:    self.assertIn('disk_usage', img)
2018-04-13 10:50:16,371.371 INFO:__main__:AssertionError: 'disk_usage' not found in {u'size': 1073741824, u'pool_name': u'rbd', u'provisioned_bytes': 0, u'name': u'cimg-clone', u'parent': {u'pool_name': u'rbd', u'image_name': u'cimg', u'snap_name': u'snap1'}, u'num_objs': 256, u'block_name_prefix': u'rbd_data.111762bbd95a', u'used_bytes': 0, u'snapshots': [], u'id': u'111762bbd95a', u'order': 22, u'timestamp': u'2018-04-13T10:50:13Z', u'features_name': [u'exclusive-lock', u'fast-diff', u'layering', u'object-map'], u'stripe_unit': 4194304, u'data_pool': None, u'stripe_count': 1, u'obj_size': 4194304, u'total_provisioned_bytes': 0, u'features': 29}
2018-04-13 10:50:16,371.371 INFO:__main__:
2018-04-13 10:50:16,371.371 INFO:__main__:----------------------------------------------------------------------
2018-04-13 10:50:16,372.372 INFO:__main__:Ran 1 test in 43.106s
2018-04-13 10:50:16,372.372 INFO:__main__:
2018-04-13 10:50:16,372.372 INFO:__main__:FAILED (failures=1)
2018-04-13 10:50:16,372.372 INFO:__main__:
2018-04-13 10:50:16,372.372 INFO:__main__:======================================================================
2018-04-13 10:50:16,372.372 INFO:__main__:FAIL: test_clone (tasks.mgr.dashboard.test_rbd.RbdTest)
2018-04-13 10:50:16,372.372 INFO:__main__:----------------------------------------------------------------------
2018-04-13 10:50:16,372.372 INFO:__main__:Traceback (most recent call last):
2018-04-13 10:50:16,372.372 INFO:__main__:  File "/ceph/qa/tasks/mgr/dashboard/test_rbd.py", line 463, in test_clone
2018-04-13 10:50:16,372.372 INFO:__main__:    'snap_name': 'snap1'})
2018-04-13 10:50:16,373.373 INFO:__main__:  File "/ceph/qa/tasks/mgr/dashboard/test_rbd.py", line 137, in _validate_image
2018-04-13 10:50:16,373.373 INFO:__main__:    self.assertIn('disk_usage', img)
2018-04-13 10:50:16,373.373 INFO:__main__:AssertionError: 'disk_usage' not found in {u'size': 1073741824, u'pool_name': u'rbd', u'provisioned_bytes': 0, u'name': u'cimg-clone', u'parent': {u'pool_name': u'rbd', u'image_name': u'cimg', u'snap_name': u'snap1'}, u'num_objs': 256, u'block_name_prefix': u'rbd_data.111762bbd95a', u'used_bytes': 0, u'snapshots': [], u'id': u'111762bbd95a', u'order': 22, u'timestamp': u'2018-04-13T10:50:13Z', u'features_name': [u'exclusive-lock', u'fast-diff', u'layering', u'object-map'], u'stripe_unit': 4194304, u'data_pool': None, u'stripe_count': 1, u'obj_size': 4194304, u'total_provisioned_bytes': 0, u'features': 29}
2018-04-13 10:50:16,373.373 INFO:__main__:
```
I'm not sure if that's really an issue or related to my environment only.

rjfd commented:
@s0nea it's not. It was a bug I introduced in the PR an hour ago, but I already fixed. Please pull the PR code again.

rjfd commented:
@sebastian-philipp thanks for the patch.

rjfd commented:
@sebastian-philipp pushed image json validation similar to your patch

dillaman commented:
@rjfd No technical reason why the async APIs aren’t exposed via Python, so definitely feel free to expand and expose.

LenzGr commented:
@sebastian-philipp have your concerns been addressed in the latest revision?

rjfd commented:
QA test results were successful after addressing the comments: http://pulpito.ceph.com/rdias-2018-04-16_13:10:22-rados:mgr-wip-rdias-testing-distro-basic-smithi/
